### PR TITLE
Added 'gelsd' to CUDA torch.linalg.lstsq driver so that underdetermined least square systems can be solved on CUDA 

### LINKED ISSUE #117122: Improve behaviour of `torch.linalg.lstsq` on CUDA GPU for rank defficient matrices
### 🚀 The feature, motivation and pitch

As per the documentation:
> For CUDA input [`torch.linalg.lstsq`] assumes that A is full-rank.

While documented, this behaviour is counter-intuitive for end-users especially if the function silently fails.

Interestingly, currently calling `torch.linalg.lstsq` on CUDA for rank defficient input silently fails in non-batched-mode but throws a `_LinAlgError` in batched mode.

It is also counter-intuitive that `torch.linalg.lstsq` on CUDA is not able to fallback to a more stable SVD driver despite `torch.linalg.svd` being supported on CUDA.

It would be great if:
1. The QR based implementation always threw an error is the input is not full rank
2. An SVD backend would be available on CUDA as well

### Alternatives

An alternative is to implement an SVD-based least-squares and use that instead of `torch.linalg.lstsq`. Here is a basic implementation (feel free to post refinements):
```python
def svd_lstsq(AA, BB, tol=1e-5):
    U, S, Vh = torch.linalg.svd(AA, full_matrices=False)
    Spinv = torch.zeros_like(S)
    Spinv[S>tol] = 1/S[S>tol]
    UhBB = U.adjoint() @ BB
    if Spinv.ndim!=UhBB.ndim:
      Spinv = Spinv.unsqueeze(-1)
    SpinvUhBB = Spinv * UhBB
    return Vh.adjoint() @ SpinvUhBB
```

### Additional context

Simple script to replicate:
```python
import torch
print(f'Running PyTorch version: {torch.__version__}')

torchdevice = torch.device('cpu')
if torch.cuda.is_available():
  torchdevice = torch.device('cuda')
  print('Default GPU is ' + torch.cuda.get_device_name(torch.device('cuda')))
print('Running on ' + str(torchdevice))

b = 2
r = 5
c = 3
k = 1

if b==1:
  A = torch.randn(r, c, device=torchdevice)
  if k==1:
    B = torch.randn(r, device=torchdevice)
  else:
    B = torch.randn(r, k, device=torchdevice)
else:
  A = torch.randn(b, r, c, device=torchdevice)
  B = torch.randn(b, r, k, device=torchdevice)

# degrade rank
A[...,-1] = 0
print("A",A)

try:
  X_lstsq = torch.linalg.lstsq(A, B).solution
  print("X_lstsq",X_lstsq)
except Exception as error:
  print("An error occurred:", type(error).__name__, "–", error)

X_pinv=torch.linalg.pinv(A) @ B
print("X_pinv",X_pinv)

def svd_lstsq(AA, BB, tol=1e-5):
    U, S, Vh = torch.linalg.svd(AA, full_matrices=False)
    Spinv = torch.zeros_like(S)
    Spinv[S>tol] = 1/S[S>tol]
    UhBB = U.adjoint() @ BB
    if Spinv.ndim!=UhBB.ndim:
      Spinv = Spinv.unsqueeze(-1)
    SpinvUhBB = Spinv * UhBB
    return Vh.adjoint() @ SpinvUhBB

X_svd= svd_lstsq(A, B)
print("X_svd",X_svd)
```

Related issues: #88101 #85021 #10454

### PR DESCRIPTION
**Pull request by cleonard530**
Fixes #117122 

Updated `torch.linalg.lstsq` to be able to use `driver='gelsd'` and solve the system of equation using the SVD of the matrix $A$ for $||Ax-b||_2^2$. This allows a user to be able to solve any system, not just the systems where $A$ has full rank. In the case where there are multiple solutions, the `'gelsd' `solver finds the solution such that $||x||_2$ is minimized. 

In the referenced issue, there is discussion about also having the 'gels' (QR solver) method always throw an error when the input is not full rank. I am not going to implement this as it would require extra work that will slow down the current algorithm. Currently, the only reason it will sometimes throw an error on batched tensors (such as the example given when the tensor is of size 2 x 5 x 3)  is because in special cases the method calls the cuBLAS `gelsBatched()` method. This method returns the `infos` parameter that tells if the matrix is rank deficient (see `lstsq_kernel()` in `aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp` in the **Files Changed**). The reasons for sometimes dispatching to cuBLAS and sometimes using the PyTorch implementation is due to performance and is discussed in this PR #54725. 

Also, note that the error for cuBLAS `gelsBatched()` does not always work. I ran the example from the linked issue but replaced:
`A[...,-1] = 0` with `A[...,-1] = A[...,-2] + A[...,-3]` 
and it gave me incorrect results without throwing any errors, so even the cuBLAS implementation `infos` argument is not always reliable. 

Currently, the `driver` argument has to be set explicitly to use `'gelsd'` but if there is convincing reason to make it a default on underdetermined systems (n_cols > n_rows) then I can add this as well.

### CODE REVIEW

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -1459,25 +1468,77 @@ void linalg_lstsq_gels(const Tensor& A, const Tensor& B, const Tensor& /*infos*/
...
+
+  ScalarType real_dtype = toRealValueType(A.scalar_type());
+
+  auto [U, S, Vh] = at::linalg_svd(A, /*full_matrices=*/false, std::nullopt);
+  singular_values.copy_(S);
+
+  Tensor tol;
+  if (rcond > 0 && rcond < 1) {
+    auto s_max = singular_values.select(-1, 0).unsqueeze(-1).expand(singular_values.sizes());
+    tol = at::mul(s_max, rcond);
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: No need to expand before `mul` -- if needed, better be done afterwards

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -1459,25 +1468,77 @@ void linalg_lstsq_gels(const Tensor& A, const Tensor& B, const Tensor& /*infos*/
...
+  } else {
+    tol = at::full_like(singular_values, _get_epsilon(real_dtype) * static_cast<double>(std::max(m, n)));
+  }
+
+  Tensor above_tol = singular_values.gt(tol);
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: No need to expand `tol` at all -- can be kept as a scalar.

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -1459,25 +1468,77 @@ void linalg_lstsq_gels(const Tensor& A, const Tensor& B, const Tensor& /*infos*/
...
+  rank.copy_((above_tol).sum(-1).to(at::kLong));
+  Tensor S_pinv = at::where(above_tol, at::reciprocal(singular_values), at::zeros_like(singular_values));
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: No need to convert to `long`. No need to create `zeros_like(singular_values)` -- `where` does broadcast. I am skeptical about doing `reciprocal` -- better use `div` and not `mul` past that. If fact, it is best to filter with `S_filtered = S * above_tol`.

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -1459,25 +1468,77 @@ void linalg_lstsq_gels(const Tensor& A, const Tensor& B, const Tensor& /*infos*/
...
+
+  auto B_narrowed = B.narrow(-2, 0, m);
+  Tensor UhB = at::matmul(U.mH(), B_narrowed);
+  Tensor Spinv_expanded = S_pinv.unsqueeze(-1).to(A.scalar_type());
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: Why calls into `to`? Isn't the SVD that of `A`?

File: test/test_linalg.py [hunk 5]
```diff
@@ -365,10 +365,10 @@ def select_if_not_empty(t, i):
...
             b_3d = b.view(batch_size, m, nrhs)
 
-            solution_3d = res.solution.view(batch_size, n, nrhs)
-            residuals_2d = apply_if_not_empty(res.residuals, lambda t: t.view(-1, nrhs))
-            rank_1d = apply_if_not_empty(res.rank, lambda t: t.view(-1))
-            singular_values_2d = res.singular_values.view(batch_size, res.singular_values.shape[-1])
+            solution_3d = res.solution.cpu().view(batch_size, n, nrhs)
+            residuals_2d = apply_if_not_empty(res.residuals.cpu(), lambda t: t.view(-1, nrhs))
+            rank_1d = apply_if_not_empty(res.rank.cpu(), lambda t: t.view(-1))
+            singular_values_2d = res.singular_values.cpu().view(batch_size, res.singular_values.shape[-1])
```
Comment (nikitaved) [inline test/test_linalg.py]: Why do we move to `cpu` here?

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp (hunk 3 above)
Comment (cleonard530) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: Hi @nikitaved, If we used `S_filtered = S * above_tol`  and used `div`, then we would have to account for the chase where `S_filtered[k] == 0` in the division. So if the matrix is not full rank we will end up with zero elements that we need to handle. One method to handle this is to do something like this:
```
  Tensor S_safe = at::where(above_tol, singular_values, 1.0);
  auto B_narrowed = B.narrow(-2, 0, m);
  Tensor UhB = at::matmul(U.mH(), B_narrowed);
  Tensor S_safe_expanded = S_safe.unsqueeze(-1).to(A.scalar_type());
  Tensor SpinvUhB = UhB.div(S_safe_expanded);
  SpinvUhB = at::where(above_tol.unsqueeze(-1), SpinvUhB, 0.0);
```
Let me know what you think about this. Once this is done I will push the rest of the changes.


Also, I will remove the .to(at::kLong)

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp (hunk 4 above)
Comment (cleonard530) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: I will remove that as well.

File: test/test_linalg.py (hunk 5 above)
Comment (cleonard530) [reply test/test_linalg.py]: This is checking that the torch.lstsq solution matches the numpy and scipy solutions. To compare the values they need to be on the same device (cpu). I believe I was getting an error without the conversion but I can check again to make sure.

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp (hunk 3 above)
Comment (nikitaved) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: We should add tests for rank-deficient cases -- these seem to be missing.

We can do `S_safe = (S * above_tol).add_(above_tal.logical_not())` -- seems to work without any explicit type promotion. This will need a comment though. (Cannot add 1 and sub `above_tol`, because `sub_(Bool Tensor)` is not supported, unfortunately).
But it is up to you, I do not have strong preferences here given that for `where` we do not need to materialize any extra memory. Let's go with `where` :) (BTW, `S_safe` can be unsqueezed in-place, and so is `above_tol`). The last `where` can be replaced with an in-place masking.

File: test/test_linalg.py (hunk 5 above)
Comment (cleonard530) [reply test/test_linalg.py]: Sorry, you're right. They do not need to be moved to cpu. I will update this as well

**Comment by nikitaved** [comment]
Sorry, @cleonard530 . I will have a look as soon as possible.

**Comment by cleonard530** [comment]
Thanks @nikitaved!

**Comment by cleonard530** [comment]
@nikitaved are you still able to take a look at this? If so, I will remove the stale tag and you can look at it when you have time, otherwise I'll leave it and let this close.
